internal/sveltos/profile.go

```diff
@@ -50,96 +51,149 @@ type HelmChartOpts struct {
 	InsecureSkipTLSVerify bool
 }
 
+// ReconcileClusterProfile reconciles a Sveltos ClusterProfile object.
+func ReconcileClusterProfile(
+	ctx context.Context,
+	cl client.Client,
+	l logr.Logger,
+	name string,
+	opts ReconcileProfileOpts,
+) (*sveltosv1beta1.ClusterProfile, error) {
+	obj := objectMeta(opts.OwnerReference)
+	obj.SetName(name)
+
+	cp := &sveltosv1beta1.ClusterProfile{
+		ObjectMeta: obj,
+	}
+
+	operation, err := ctrl.CreateOrUpdate(ctx, cl, cp, func() error {
+		spec, err := Spec(&opts)
+		if err != nil {
+			return err
+		}
+		cp.Spec = *spec
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if operation == controllerutil.OperationResultCreated || operation == controllerutil.OperationResultUpdated {
+		l.Info(fmt.Sprintf("Successfully %s ClusterProfile %s", string(operation), cp.Name))
+	}
+
+	return cp, nil
+}
+
 // ReconcileProfile reconciles a Sveltos Profile object.
-func ReconcileProfile(ctx context.Context,
+func ReconcileProfile(
+	ctx context.Context,
 	cl client.Client,
 	l logr.Logger,
 	namespace string,
 	name string,
-	matchLabels map[string]string,
 	opts ReconcileProfileOpts,
 ) (*sveltosv1beta1.Profile, error) {
-	cp := &sveltosv1beta1.Profile{
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: namespace,
-			Name:      name,
-		},
+	obj := objectMeta(opts.OwnerReference)
+	obj.SetNamespace(namespace)
+	obj.SetName(name)
+
+	p := &sveltosv1beta1.Profile{
+		ObjectMeta: obj,
 	}
 
+	operation, err := ctrl.CreateOrUpdate(ctx, cl, p, func() error {
+		spec, err := Spec(&opts)
+		if err != nil {
+			return err
+		}
+		p.Spec = *spec
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if operation == controllerutil.OperationResultCreated || operation == controllerutil.OperationResultUpdated {
+		l.Info(fmt.Sprintf("Successfully %s Profile %s", string(operation), p.Name))
+	}
+
+	return p, nil
+}
+
+// Spec returns a spec object to be used with
+// a Sveltos Profle or ClusterProfile object.
+func Spec(opts *ReconcileProfileOpts) (*sveltosv1beta1.Spec, error) {
 	tier, err := PriorityToTier(opts.Priority)
 	if err != nil {
 		return nil, err
 	}
 
-	operation, err := ctrl.CreateOrUpdate(ctx, cl, cp, func() error {
-		if cp.Labels == nil {
-			cp.Labels = make(map[string]string)
-		}
+	spec := &sveltosv1beta1.Spec{
+		ClusterSelector: libsveltosv1beta1.Selector{
+			LabelSelector: opts.LabelSelector,
+		},
+		Tier:               tier,
+		ContinueOnConflict: !opts.StopOnConflict,
```

Comment:
can we preallocate the slice here like `HelmCharts: make([]sveltosv1beta1.HelmChart, 0, len(opts.HelmChartOpts)),`?